### PR TITLE
Fix memory leak bug in build_timeout_circuit

### PR DESCRIPTION
**Pull request by david415**

### CODE REVIEW

File: txtorcon/circuit.py [hunk 1]
```diff
@@ -289,12 +289,23 @@ def build_timeout_circuit(tor_state, reactor, path, timeout, using_guards=False)
...
     receive a circuit build result within the `timeout` duration.
     """
+    timed_circuit = None
     d = tor_state.build_circuit(path, using_guards)
-    reactor.callLater(timeout, d.cancel)
 
-    def trap_cancel(f):
+    def get_circuit(c):
+        timed_circuit = c
+        return c
```
Comment (meejah) [inline txtorcon/circuit.py]: I think what you intend here isn't supported by Python (i.e. you want to assign to the outer-scope `timed_circuit`, but this assignment makes a local-only assignment to `timed_circuit` so I don't believe you can ever hit the "`is not None`" clause in the errback...

**Comment by david415** [comment]
i'm not sure how to get the unit tests to exercise the cancellation code path where we have the Circuit object.

**Comment by meejah** [comment]
You should be able to pass a Twisted `Clock()` instance as the reactor, and then `.advance()` it past your timeout? For an example of this, see the `AddrMap` tests which test addresses expiring.

**Comment by meejah** [comment]
Okay, so it wasn't quite as simple as a `Clock` but see my PR with a test -- the extra bit being that we needed a test for a partially-built circuit that reached the timeout.

**Comment by david415** [comment]
yes! thanks! that fixed it.

**Pull request merged**